quickwit/quickwit-indexing/src/actors/cooperative_indexing.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::hash::{DefaultHasher, Hasher};
+use std::sync::Arc;
+use std::time::Duration;
+
+use quickwit_proto::indexing::{CpuCapacity, PipelineMetrics, PIPELINE_FULL_CAPACITY};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+use tokio::time::Instant;
+
+const NUDGE_TOLERANCE: Duration = Duration::from_secs(2);
+
+pub struct CooperativeIndexing {
+    target_phase: Duration,
+    commit_timeout: Duration,
+    shared_semaphore: Arc<Semaphore>,
+    t0: Instant,
+}
+
+impl CooperativeIndexing {
+    pub fn new(
+        id: impl std::hash::Hash,
+        commit_timeout: Duration,
+        shared_semaphore: Arc<Semaphore>,
+    ) -> CooperativeIndexing {
+        assert!(commit_timeout.as_millis() > 0);
+        let mut hasher = DefaultHasher::new();
+        id.hash(&mut hasher);
+        let target_phase_ms: u64 = hasher.finish() % commit_timeout.as_millis() as u64;
```

Comment:
I think `_millis` is more readable. Trinty also has a PR with many `_s` and `_ms` and I found it hard to read.